ssh/tailssh/child.go

```diff
@@ -0,0 +1,246 @@
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+//go:build linux || (darwin && !ios)
+// +build linux darwin,!ios
+
+package tailssh
+
+import (
+	"context"
+	"fmt"
+	"io"
+	"log"
+	"log/syslog"
+	"os"
+	"os/exec"
+	"strconv"
+	"strings"
+	"syscall"
+	"unsafe"
+
+	"github.com/creack/pty"
+	"github.com/gliderlabs/ssh"
+	"tailscale.com/cmd/tailscaled/childproc"
+	"tailscale.com/types/logger"
+)
+
+func init() {
+	childproc.Add("ssh", sshChild)
+}
+
+var ptyName = func(f *os.File) (string, error) {
+	return "", fmt.Errorf("unimplemented")
+}
+
+var startLoginSession = func(logf logger.Logf, uid uint32, localUser, remoteUser, remoteHost, tty string) (close func() error, err error) {
+	return nil, fmt.Errorf("unimplemented")
+}
+
+// newSSHChildCommand returns a new exec.Cmd configured with
+// `tailscaled be-child ssh` as the entrypoint.
+// If tailscaled is empty, the subcommand is not configured.
+func newSSHChildCommand(ctx context.Context, uid string, ci *sshConnInfo, tailscaled, name string, args []string) *exec.Cmd {
+	if tailscaled != "" {
+		args = append([]string{"be-child", "ssh", name}, args...)
+		name = tailscaled
+	}
+
+	cmd := exec.CommandContext(ctx, name, args...)
+	if tailscaled != "" {
+		// Don't bother setting these if we aren't spawning tailscaled.
+		cmd.Env = []string{
+			"TS_UID=" + uid,
+			"TS_LOCAL_USER=" + ci.localUser,
+			"TS_REMOTE_USER=" + ci.uprof.LoginName,
+			"TS_REMOTE_TAGS=" + strings.Join(ci.node.Tags, ","),
+			"TS_REMOTE_IP=" + ci.src.IP().String(),
+		}
+	}
+	return cmd
+}
+
+const debugChild = false
+
+// sshChild is the entrypoint to the `tailscaled be-child ssh` subcommand.
+// This adds a necessary layer of indirection between tailscaled and the
+// process that we are about to spawn. One of the primary requirements
+// is to inform the system of new login session from the user. This is
+// typically necessary for mounting home directories and decrypting file
+// systems.
+// It expects the following env variables:
+// - TS_UID
+// - TS_LOCAL_USER
+// - TS_REMOTE_USER
+// - TS_REMOTE_TAGS
+// - TS_REMOTE_IP
+// - TS_HAS_TTY
+// - TS_TTY_NAME
+func sshChild(args []string) error {
+	if len(args) < 1 {
+		return fmt.Errorf("no args provided")
+	}
+	logf := logger.Discard
+	if debugChild {
+		// The only place we can log is syslog.
+		if sl, err := syslog.New(syslog.LOG_INFO|syslog.LOG_DAEMON, "tailscaled-ssh"); err == nil {
+			logf = log.New(sl, "", 0).Printf
+		}
+	}
+	localUser := os.Getenv("TS_LOCAL_USER")
+
+	uid, err := strconv.ParseUint(os.Getenv("TS_UID"), 10, 32)
+	if err != nil {
+		logf("ssh: user Lookup %q: %v", localUser, err)
+		os.Exit(1)
+	}
+
+	euid := uint64(os.Geteuid())
+	if euid == 0 {
+		// Inform the system that we are about to login someone.
+		// We can only do this if we are running as root.
+		remoteUser := os.Getenv("TS_REMOTE_TAGS")
+		if remoteUser == "" {
+			remoteUser = os.Getenv("TS_REMOTE_USER")
```

Comment:
using env vars for this scares me for security reasons.

people can ssh -E and send malicious stuff here maybe?